### DISCUSSION TITLE
Wrapping a ListStore with an Rc no longer (?) works anymore

### DISCUSSION THREAD

**Discussion by cmdcolin** [gtk-rs/gtk4-rs#1301] [Q&A] [answered]
I was trying to update a sample app that worked with 0.5.x using a ListStore in a Rc. Example code similar to the column_view_datagrid example

setup
```

    let store = gio::ListStore::new(BoxedAnyObject::static_type());

    (0..10000).for_each(|i| {
        store.append(&BoxedAnyObject::new(Row {
            col1: format!("col1 {i}"),
            col2: format!("col2 {i}"),
        }))
    });
```

fails with compile error

```
    let store_rc = Rc::new(store);
    let sel = gtk::SingleSelection::new(Some(&store_rc)); // fails with compile error
    let columnview = gtk::ColumnView::new(Some(&sel))

```

works

```
    let sel = gtk::SingleSelection::new(Some(&store)); // works
    let columnview = gtk::ColumnView::new(Some(&sel))
```

the compile error is

```
error[E0277]: the trait bound `Rc<gtk4::gio::ListStore>: IsA<ListModel>` is not satisfied
  --> examples/column_view_datagrid/main.rs:44:41
   |
44 |     let sel = gtk::SingleSelection::new(Some(&store_rc));
   |               ------------------------- ^^^^^^^^^^^^^^^ the trait `IsA<ListModel>` is not implemented for `Rc<gtk4::gio::ListStore>`
   |               |
   |               required by a bound introduced by this call
   |
   = help: the following other types implement trait `IsA<T>`:
             <ATContext as IsA<ATContext>>
             <ATContext as IsA<gtk4::glib::Object>>
             <AboutDialog as IsA<AboutDialog>>
             <AboutDialog as IsA<Accessible>>
             <AboutDialog as IsA<Buildable>>
             <AboutDialog as IsA<ConstraintTarget>>
             <AboutDialog as IsA<Native>>
             <AboutDialog as IsA<Root>>
           and 1763 others
note: required by a bound in `SingleSelection::new`
  --> /home/cdiesh/src/gtk4-rs/gtk4/src/auto/single_selection.rs:29:36
   |
29 |     pub fn new(model: Option<&impl IsA<gio::ListModel>>) -> SingleSelection {
   |                                    ^^^^^^^^^^^^^^^^^^^ required by this bound in `SingleSelection::new`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `gtk4-rs-examples` due to previous error


```

this used to work for me but after i tried updating to 0.6.0 started giving the above error. 

here is an example PR for attempting the 0.6.0 upgrade that producing this error

https://github.com/cmdcolin/fml9000/pull/17/files

trying rustc 1.69.0-nightly (2773383a3 2023-02-10) but also was affecting rustc 1.68.0-nightly (77429957a 2023-01-01)

**Comment by sdroege** [comment, marked as answer]
Why do you want to wrap it in an `Rc`? The `ListStore` already behaves like an `Rc`: `clone()` will simply increase its reference count.

For your code, it will probably work if you do `store_rc.as_ref()` but having two layers of reference counting is not a good idea.

**Reply by cmdcolin** [reply]
thanks i think i fixed it without using an Rc now. that Rc was probably unneeded, i'm quite new to rust and still have a pretty weak grasp on the memory management and sometimes wrap things in Rc just to avoid some of the compiler complaints

I managed to fix the PR using just clone and I think it is working again

here is my updated code if there is interest, quite naive approaches all around in the code probably

https://github.com/cmdcolin/fml9000/pull/17

**Reply by sdroege** [reply]
Seems fine to me. You might be able to get rid of some of the `clone()`s depending on the surrounding code but that doesn't really matter much for reference counting.

Knowing whether a type in Rust does deep copies or refcounting or something else is indeed not obvious.
